[RFC] fix incorrect GSI numbering causing device limit on aarch64

Swchexit opened the pull request:
I’m opening this as an RFC because I would appreciate feedback on my proposed solution. Please see my detailed analysis and reasoning on the issue page #4207 . Additionally, I’m encountering timeout problems running the test suite with my current setup and would appreciate assistance verifying this patch.

## Changes
Introduce two new constants (GSI_BASE and GSI_MAX) in the layout.rs files for both aarch64 and x86_64. These constants correctly adjust the GSI numbering for aarch64 VMs, aligning Firecracker’s interrupt assignments with Linux’s expectations and ensuring that the interrupts provided in the device tree and registered via irqfd are correctly numbered.

## Reason
To resolve issue #4207 , where incorrect GSI numbering caused by using an offset of 32 (IRQ_BASE) resulted in an artificial limitation of 64 devices on aarch64 microVMs. This change removes that limitation, enabling aarch64 microVMs to correctly use up to 95 devices.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] I have read and understand [CONTRIBUTING.md][3].
- [x] I have run `tools/devtool checkstyle` to verify that the PR passes the
  automated style checks.
- [x] I have described what is done in these changes, why they are needed, and
  how they are solving the problem in a clear and encompassing way.
- [x] I have updated any relevant documentation (both in code and in the docs)
  in the PR.
- [ ] ~~I have mentioned all user-facing changes in `CHANGELOG.md`.~~
- [X] If a specific issue led to this PR, this PR closes the issue.
- [ ] ~~When making API changes, I have followed the
  [Runbook for Firecracker API changes][2].~~
- [ ] I have tested all new and changed functionalities in unit tests and/or
  integration tests.
- [ ] ~~I have linked an issue to every new `TODO`.~~

______________________________________________________________________

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

Swchexit commented on src/vmm/src/arch/aarch64/layout.rs:
```diff
@@ -80,5 +80,15 @@ pub const IRQ_MAX: u32 = 128;
 
+// The Linux kernel automatically shifts the GSI by 32 if it is an SPI,
+// allowing us to start numbering from 0 instead of 32.
+// But device_manager/mmio.rs states that 0 is not allowed for GSIs
+// So we start with 1 instead.
```
Yeah you're right. Changed that to start from 0 and works fine. I'll make sure to update this soon.

Manciukic replied:
Looking at the PR that introduced the `NonZeroU32` (#4601), I don't see why it was done, honestly. `KVM_IRQFD` ioctl has no documented constraints, and looking at aarch64 code, it seems that 0 is valid (as it's mapped to SPI 32). On x86 `IRQ_BASE` will already ensure it's non-zero, but it seems "valid" from the ioctl point of view.

Swchexit replied:
Hi @roypat, thanks for the explanation. I have thus updated the commit to start allocating GSIs from zero, and now we can use 96 devices fully. This changed the type of `MMIODeviceInfo::irq` from `Option<NonZeroU32>` to `Option<u32>`. How do you think about this patch? Sorry that I'm new to this project, and I'm unsure what's the next step for me.

roypat replied:
The rust changes look absolutely fine to me. It'd be cool if we could extend the `test_max_devices.py` test to also cover ARM now, and assert that we can actually attach and use 96 devices.

Swchexit replied:
Update: the tests have been updated, and now they can pass reliably. The total usable devices would be `94` because `vmgenid` and `pl301 rtc` devices are non-optional and each takes one interrupt line.

Swchexit commented:
Hi, I've fixed the clippy error, updated the CHANGELOG, and removed the description of the 64-device limit from the docs. I will try to run the extended test as soon as I fix my settings.

roypat commented:
Alright, I think the last failures we can fix by just disabling the memory monitor in the "too many devices" test (pass `monitor_memory=False` to `microvm_factory.build()`).

Manciukic commented:
Thanks for the contribution!
